app/controllers/api/v0/versions_controller.rb

```diff
@@ -25,6 +27,41 @@ def show
     }
   end
 
+  def raw
+
+    @version ||= version_collection.find(params[:id])
+
+    def self.public_hosts=(hosts)
+      hosts = [] if hosts.nil?
+      hosts = hosts.split(' ') if hosts.is_a?(String)
+      unless hosts.is_a?(Enumerable) && hosts.all? {|host| host.is_a?(String)}
+        raise StandardError, 'Public hosts must be a string or enumerable of strings'
+      end
+      @public_hosts = hosts
+    end
+
+    if ENV['PUBLIC_ARCHIVE_HOSTS']
+      self.public_hosts = ENV['PUBLIC_ARCHIVE_HOSTS']
+    end
+
+    def self.public_hosts
+      @public_hosts || []
+    end
+
+    def self.public_archive_uri?(uri)
+      public_hosts.any? {|base| uri.starts_with?(base)}
+    end
+
+    if public_archive_uri?(@version.uri)
+      redirect_to @version.uri, status: 301 and return
+    else
+      upstream = open(@version.uri)
+      mime_type = @version.source_metadata['mime_type']
+      mime_type = upstream.content_type if mime_type.nil? || mime_type.empty?
+      send_data(upstream.read, type: mime_type, disposition: 'inline')
```

Comment:
Should probably set some long cache headers here, too.

Reply:
Done. Redirects are now cacheable as well. Is that desirable?

Reply:
I don’t think there should be a problem with that. That said, I’m not even sure if most clients respect cache headers on redirects (or how they cross them with what different 3xx codes mean if they do respect them). Are you concerned that we might decide in the future to move everything sometime and this would cause us trouble?

Reply:
I mostly just checking in, because I noticed that firefox will recheck the original URL on a hard reload if the redirect wasn't cacheable, but will just go straight to the 301 target if the redirect was cacheable. I imagine that the behavior with the current response varies quite a bit in different user agents. If we do end up having an unforeseen move of assets from the current s3 buckets, we can always leave the current bucket up with redirects.

Reply:
> Firefox will recheck the original URL on a hard reload if the redirect wasn't cacheable, but will just go straight to the 301 target if the redirect was cacheable.

That’s fascinating and good to know!

> I mostly just checking in… If we do end up having an unforeseen move of assets from the current s3 buckets, we can always leave the current bucket up with redirects.

Cool, so I think I think we’re good here. 👍